scripts/azureml-assets/azureml/assets/config.py

```diff
@@ -582,49 +602,76 @@ def uri(self) -> str:
         # do not exist then fail gracefully, return the URI "as-is", and hope for the best.
         try:
             blob_service_client = BlobServiceClient(
-                account_url=self._account_uri,
+                account_url=account_uri,
                 credential=AzureCliCredential(
                     process_timeout=AzureBlobstoreAssetPath.AZURE_CLI_PROCESS_LOGIN_TIMEOUT
                 )
             )
-            container_client = blob_service_client.get_container_client(container=self._container_name)
+            container_client = blob_service_client.get_container_client(container=container_name)
 
             # If the container allows for anonymous access then we can return the URI "as-is"
             if container_client.get_container_properties().public_access is not None:
-                return self._uri
+                return uri
 
             # Our final approach is to generate a SAS token for the container and append
             # it to the URI
-            start_time = datetime.datetime.now(datetime.timezone.utc)
-            expiry_time = start_time + AzureBlobstoreAssetPath.SAS_EXPIRATION_TIME_DELTA
-
             key = blob_service_client.get_user_delegation_key(start_time, expiry_time)
 
             sas_token = "?" + generate_container_sas(
-                account_name=self._storage_name,
-                container_name=self._container_name,
+                account_name=storage_name,
+                container_name=container_name,
                 user_delegation_key=key,
                 permission=ContainerSasPermissions(read=True, list=True),
                 expiry=expiry_time,
                 start=start_time
             )
 
-            self._uri += sas_token
-            return self._uri
+            uri += sas_token
+            return uri
 
         except Exception:
             # If we fail then simply return the URI "as-is" and hope for the best
+            return uri
+
+    @property
+    def uri(self) -> str:
+        """Asset URI.""" 
+        # If we have already cached the URI, then simply return it
+        if self._uri is not None:
+            return self._uri
+        
+        # Build the simple, non-SAS URI for quick use later in the function.
+        # Its possible that the account URI may need additional tweaking to add a SAS
+        # token if the account does not allow for anonymous access.
+        self._uri = f"{self._account_uri}/{self._container_name}/{self._container_path}"
+        
+        if self.generate_uri(self._uri, self._account_uri, self._container_name, self._token) is not None:
             return self._uri
+        
+        start_time = datetime.datetime.now(datetime.timezone.utc)
+        expiry_time = start_time + AzureBlobstoreAssetPath.SAS_EXPIRATION_TIME_DELTA
+
+        return self.generate_sas_token(self._uri, self._account_uri, self._container_name, self._storage_name, start_time, expiry_time)

```

Comment:
I think it would help to swap this with line 648 above. Generate the SAS token first, if not already done, and then create the URI using the account URI and token.